geomfn: implement the majority of the indexed geom builtins

otan opened the pull request:
This commit adds builtins for the majority of geometry indexed backed
functions by calling into GEOS. The missing ones are `DWithin` and
`DFullyWithin`, which are not backed by GEOS.

Added some basic unit tests that don't do too much as this should be
tested by GEOS. We have not linked this to the builtins yet.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47495)
<!-- Reviewable:end -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47495#-:-M4ytteiBq5XjrZ0y7hg:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geos/geos_unix.cc, line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47495#-M4yry-P-eT9lbfgd3cU:-M4yry-Q83X8uPZGYCYI:biv22t3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bd0c2554ada58e183ec8edf4242d83fe490c947/pkg/geo/geos/geos_unix.cc#L274)):*
> ```cpp
> //
> 
> template <typename T>
> ```

can you change the `char* ret` to `bool* ret` -- it makes things clearer.
___
*[pkg/geo/geos/geos_unix.go, line 241 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47495#-M4yt52j7Vl-i1j3XhNk:-M4yt52j7Vl-i1j3XhNl:b-sxu2px) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bd0c2554ada58e183ec8edf4242d83fe490c947/pkg/geo/geos/geos_unix.go#L241)):*
> ```Go
> }
> 
> // Crosses returns whether the EWKB provided by A is covered by the EWKB provided by B.
> ```

s/is covered by/spatially crosses/

And can we add a link in this file to the canonical definition of these. I lookup `ST_Crosses` when seeing this, but a passing reader can't necessarily do that correlation.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47495)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geos/geos_unix.cc, line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47495#-M4yry-P-eT9lbfgd3cU:-M4z6KsJ4wI4wWdGmjuH:b-xyx9pb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bd0c2554ada58e183ec8edf4242d83fe490c947/pkg/geo/geos/geos_unix.cc#L274)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you change the `char* ret` to `bool* ret` -- it makes things clearer.
</blockquote></details>

i can't seem to because the C bindings have `char` and `bool` doesn't exist in C IIRC.
___
*[pkg/geo/geos/geos_unix.go, line 241 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47495#-M4yt52j7Vl-i1j3XhNk:-M4z6gYa4tRToRbKnSEe:b-5nr2uv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bd0c2554ada58e183ec8edf4242d83fe490c947/pkg/geo/geos/geos_unix.go#L241)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

s/is covered by/spatially crosses/

And can we add a link in this file to the canonical definition of these. I lookup `ST_Crosses` when seeing this, but a passing reader can't necessarily do that correlation.
</blockquote></details>

there isn't anything to hyperlink against :( unless you had something in mind.


<!-- Sent from Reviewable.io -->

otan committed:
geomfn: implement the majority of the indexed geom builtins

This commit adds builtins for the majority of geometry indexed backed
functions by calling into GEOS. The missing ones are `DWithin` and
`DFullyWithin`, which are not backed by GEOS.

Added some basic unit tests that don't do too much as this should be
tested by GEOS. We have not linked this to the builtins yet.

Release note: None

otan commented:
can update in follow up :)

bors r=sumeerbhola
